pkg/service/controller.go

```diff
@@ -170,10 +169,38 @@ func (c *ControllerService) CreateVolume(ctx context.Context, req *csi.CreateVol
 				busParameter:    string(bus),
 				serialParameter: serial,
 			},
+			ContentSource: req.GetVolumeContentSource(),
 		},
 	}, nil
 }
 
+func (c *ControllerService) determineDvSource(ctx context.Context, req *csi.CreateVolumeRequest) (*cdiv1.DataVolumeSource, error) {
+	res := &cdiv1.DataVolumeSource{}
+	if req.GetVolumeContentSource() != nil {
+		source := req.GetVolumeContentSource()
+		switch source.Type.(type) {
+		case *csi.VolumeContentSource_Snapshot:
+			if snapshot, err := c.virtClient.GetVolumeSnapshot(ctx, c.infraClusterNamespace, source.GetSnapshot().GetSnapshotId()); errors.IsNotFound(err) {
+				return nil, status.Errorf(codes.NotFound, "source snapshot content %s not found", source.GetSnapshot().GetSnapshotId())
+			} else if err != nil {
+				return nil, err
+			} else if snapshot != nil {
+				if snapshotSource := source.GetSnapshot(); snapshotSource != nil {
+					res.Snapshot = &cdiv1.DataVolumeSourceSnapshot{
+						Name:      snapshot.Name,
+						Namespace: c.infraClusterNamespace,
+					}
```

Comment:
Are you sure we want to use the snapshot source?  It does a lot of extra stuff like expand for free space and support for different source/target volume modes.  Thinking it may be better to simply be a passthrough to whatever the underlying CSI driver supports.  So instead of setting `spec.sourc.snapshot`.  Maybe go with `spec.pvc.dataSourceRef`?

Reply:
If DataVolumes are used in tenant will free space expansion happen twice?

Reply:
I can understand the desire to have free space applied (once) to new volumes but I am less clear how it is useful/desirable in PVC from snapshot case.  The snapshot already reports how big it is (including used overhead) and users can base their desired size off of that

Reply:
But the users only see the size of the snapshot in the tenant cluster. If they restore that into a PVC in the tenant cluster, then the restored PVC in the infra cluster would need fs overhead if it is based on a filesystem.